squint.c

```diff
@@ -0,0 +1,1611 @@
+         info &= ~RI_RmAct;
+      }
+
+      /* keep register Ids for active registers */
+      info |= inst & activeRegMask[(info & RI_Active) >> 4];
+
+      *rInfo++ = info;
+   }
+
+   /* termination sentinel to simplify reg_info scans */
+   *rInfo = 0xffffffff;
+
+   /* fixup: bump basic block begin markers forward by one instruction */
+   for (scan = rInfo-1; scan >= instInfo; --scan) {
+      if (*(scan-1) & RI_bb) {
+         *scan     |= RI_bb;
+         *(scan-1) ^= RI_bb;
+      }
+   }
+}
+
+/* find first def of reg in given direction */
+int *find_def(int *instInfo, int *rInfo, int reg, int direction)
+{
+   int *retVal = 0;
+   int info;
+
+   while( (rInfo != instInfo) &&
+          *rInfo != 0xffffffff ) { /* if in-bounds of func */
+      info = *rInfo;
+      if (info & RI_hasD) {
+         int cmpReg = (info & RI_RnDest) ?
+                      (info & RI_Rn) >> 16 : (info & RI_Rd) >> 12;
+         if (reg == cmpReg) {
+            retVal = rInfo;
+            break;
+         }
+      }
+      else if ((info & RI_func) && reg == 0) {
+         retVal = rInfo; // only good for location, not content
+         break;
+      }
+      rInfo += direction;
+   }
+
+   return retVal;
+}
+
+/* find first use of reg in given direction */
+int *find_use(int *instInfo, int *rInfo, int reg, int direction)
+{
+   int *retVal = 0;
+   int info;
+
+   while( (rInfo != instInfo) &&
+          *rInfo != 0xffffffff ) { /* if in-bounds of func */
+      info = *rInfo;
+      if (info & RI_hasD) {
+         // if not a memory op or a memory read, disable dest reg
+         if ((info & RI_memOp) == 0 || (info & RI_memRW)) {
+            info ^= (info & RI_RnDest) ? RI_RnAct : RI_RdAct;
+         }
+      }
+      if ((info & RI_RmAct) && (info & RI_Rm) == reg) {
+         retVal = rInfo;
+         break;
+      }
+      if (info & RI_RsAct && ((info & RI_Rs)>>8) == reg) {
+         retVal = rInfo;
+         break;
+      }
+      if (info & RI_RdAct && ((info & RI_Rd)>>12) == reg) {
+         retVal = rInfo;
+         break;
+      }
+      if (info & RI_RnAct && ((info & RI_Rn)>>16) == reg) {
+         retVal = rInfo;
+         break;
+      }
+      rInfo += direction;
+   }
+
+   return retVal;
+}
+
+/**********************************************************/
+/********** peephole optimization funcs *******************/
+/**********************************************************/
+
+void apply_peepholes1(int *funcBegin, int *funcEnd)
+{
+   int *scan;
+   int *scanp1, *scanp2, *scanp3;
+   for (scan = funcBegin; scan < funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+      /* one instruction peepholes */
+      if (*scan == 0xe2400000 || /* sub r0, r0, #0 */
+          *scan == 0xe2800000) { /* add r0, r0, #0 */
+         *scan = NOP;
+      }
+      else {    /* two instruction peephole */
+
+         scanp1 = active_inst(scan, 1);
+
+         if ((*scan & 0xfff00000) == 0xe2800000 || // add rI, rS, #X
```

Comment:
Should you check `(*scan & 0xfff00f00)` instead of `(*scan & 0xfff00000)`?

According to ARM reference manual, [ADD (immediate)](https://developer.arm.com/documentation/ddi0406/cb/Application-Level-Architecture/Instruction-Details/Alphabetical-list-of-instructions/ADD--immediate--ARM-?lang=en) uses [Modified immediate constants](https://developer.arm.com/documentation/ddi0406/cb/Application-Level-Architecture/ARM-Instruction-Set-Encoding/Data-processing-and-miscellaneous-instructions/Modified-immediate-constants-in-ARM-instructions?lang=en), which splits 12-bit immediate into upper 4-bit for rotation and lower 8-bit for the value (to be rotated).  OTOH, [LDR (immediate)](https://developer.arm.com/documentation/ddi0406/cb/Application-Level-Architecture/Instruction-Details/Alphabetical-list-of-instructions/LDR--immediate--ARM-?lang=en) simply zero extend the 12-bit.

Reply:
Good eye. I will check.